modules/openbsd/openbsd-driver.c

```diff
@@ -0,0 +1,135 @@
+  log_reader_options_destroy(&self->reader_options);
+  log_src_driver_free(s);
+}
+
+static gboolean
+_openbsd_sd_init(LogPipe *s)
+{
+  OpenBSDDriver *self = (OpenBSDDriver *)s;
+  GlobalConfig *cfg = log_pipe_get_config(s);
+
+  if (!log_src_driver_init_method(s))
+    return FALSE;
+
+  log_reader_options_init(&self->reader_options, cfg, self->super.super.group);
+
+  if (socketpair(AF_UNIX, SOCK_STREAM, PF_UNSPEC, self->pair) == -1)
+    {
+      msg_error("openBSD source: cannot bind socket fd");
+      return FALSE;
+    }
+
+  if ((self->klog = open(self->dev_filename->str, O_RDONLY, 0)) == -1)
+    {
+      msg_error("openBSD source: cannot open log device",
+                evt_tag_str(EVT_TAG_FILENAME, self->dev_filename->str),
+                evt_tag_errno(EVT_TAG_OSERROR, errno));
+      return FALSE;
+    }
+
+  if (ioctl(self->klog, LIOCSFD, &(self->pair[1])) == -1)
+    {
+      msg_error("openBSD source: cannot do ioctl LIOCSFD");
+      return FALSE;
+    }
+
+  close(self->pair[1]);
+  self->reader = log_reader_new(cfg);
+  log_reader_reopen(self->reader, log_proto_dgram_server_new(log_transport_stream_socket_new(self->pair[0]),
+                                                             &self->reader_options.proto_options.super),
+                    poll_fd_events_new(self->pair[0]));
+
+  log_reader_set_options(self->reader,
+                         s,
+                         &self->reader_options,
+                         self->super.super.id,
+                         self->dev_filename->str);
+  log_pipe_append((LogPipe *) self->reader, s);
+
+  if (!log_pipe_init((LogPipe *) self->reader))
+    {
+      msg_error("Error initializing log_reader, closing fd",
+                evt_tag_int("fd", self->pair[0]));
+      log_pipe_unref((LogPipe *) self->reader);
+      self->reader = NULL;
+      close(self->pair[0]);
+      close(self->klog);
+      return FALSE;
+    }
+
+  return TRUE;
+}
+
+static gboolean
+_openbsd_sd_deinit(LogPipe *s)
+{
+  OpenBSDDriver *self = (OpenBSDDriver *)s;
+
+  if (self->reader)
+    {
+      log_pipe_deinit((LogPipe *) self->reader);
+      log_pipe_unref((LogPipe *) self->reader);
+      self->reader = NULL;
+    }
+
+  if (!log_src_driver_deinit_method(s))
+    return FALSE;
+
+
+  return TRUE;
+}
+
+LogDriver *
+openbsd_sd_new(gchar *filename,GlobalConfig *cfg)
+{
+  OpenBSDDriver *self = g_new0(OpenBSDDriver, 1);
+
+  log_src_driver_init_instance(&self->super, cfg);
+
+  self->dev_filename = g_string_new(filename);
+  self->super.super.super.init    = _openbsd_sd_init;
+  self->super.super.super.deinit  = _openbsd_sd_deinit;
+  self->super.super.super.free_fn = _openbsd_sd_free;
+  log_reader_options_defaults(&self->reader_options);
+  self->reader_options.parse_options.flags |= LP_LOCAL;
+  self->reader_options.parse_options.flags &= ~LP_EXPECT_HOSTNAME;
+  self->reader_options.super.stats_level = STATS_LEVEL1;
+  self->reader_options.super.stats_source = SCS_UNIX_STREAM;
+
```

Comment:
We might want to use a separate stats source.

Reply:
I defined a new stats source:
SCS_OPENBSD  = 42
Is it the correct approach or shall I use one of the existing SCS_ ?